Added a configuration option on how to label PCR data

Mailaender opened the pull request:
Closes https://github.com/eclipse/chemclipse/issues/661

Mailaender committed:
Use the channel name for labels and tooltips.

eselmeister commented on chemclipse/plugins/org.eclipse.chemclipse.ux.extension.xxd.ui/src/org/eclipse/chemclipse/ux/extension/xxd/ui/swt/ExtendedPlateChartsUI.java:
```diff
@@ -271,6 +273,18 @@ private void updateChart() {
+	private String getLabel(IWell well) {
+
+		LabelSetting referenceLabel = LabelSetting.valueOf(preferenceStore.getString(PreferenceConstants.P_PCR_REFERENCE_LABEL));
+		if(referenceLabel.equals(LabelSetting.SAMPLENAME))
```
Please add curly braces. Are the formatting settings set in your IDE?

Mailaender replied:
I enabled the `Use blocks in if/while/for/do statements` save action now.

eselmeister commented on chemclipse/plugins/org.eclipse.chemclipse.ux.extension.xxd.ui/src/org/eclipse/chemclipse/ux/extension/xxd/ui/preferences/PreferenceConstants.java:
```diff
@@ -778,6 +779,8 @@
 	public static final String P_PCR_COLOR_CODES = "pcrColorCodes";
 	public static final String DEF_PCR_COLOR_CODES = "";
+	public static final String P_PCR_REFERENCE_LABEL = "pcrReferenceLabel";
+	public static final String DEF_PCR_REFERENCE_LABEL = LabelSetting.COORDINATE_SAMPLENAME.getLabel();
```
I would recommend to persist "LabelSetting.COORDINATE_SAMPLENAME.name()". The label itself if variable and could be translated in the future. The enum entry name itself is independent of a possibly translated label.

eselmeister commented on chemclipse/plugins/org.eclipse.chemclipse.ux.extension.xxd.ui/src/org/eclipse/chemclipse/ux/extension/xxd/ui/swt/ExtendedPlateChartsUI.java:
```diff
@@ -271,6 +273,18 @@ private void updateChart() {
 
+	private String getLabel(IWell well) {
+
+		LabelSetting referenceLabel = LabelSetting.valueOf(preferenceStore.getString(PreferenceConstants.P_PCR_REFERENCE_LABEL));
```
Handle: IllegalArgumentException | NullPointerException

Mailaender replied:
I don't see them handled anywhere, so I will just add a default fallback here.

Mailaender commented:
Updated.

eselmeister commented:
What happens if 

`LabelSetting referenceLabel = LabelSetting.valueOf(preferenceStore.getString(PreferenceConstants.P_PCR_REFERENCE_LABEL));`

the return of `preferenceStore.getString(PreferenceConstants.P_PCR_REFERENCE_LABEL)` is null or "" or doesn't match any enum name?

Mailaender commented:
```java
	private String getString(Properties p, String name) {
		String value = p != null ? p.getProperty(name) : null;
		if (value == null) {
			return STRING_DEFAULT_DEFAULT;
		}
		return value;
	}
```

will never return null. `STRING_DEFAULT_DEFAULT` is an empty string. I added a fallback to the default value instead of no label. Also added enum parsing that will not crash the widget, but also go to a fallback.

I now also set the ID of line series and use the description field instead.

eselmeister commented:
Looks good. Two comments:

A) When introducing a new dependency, I recommend to also set the required version in the MANIFEST.MF:
`com.google.guava;bundle-version="27.1.0"`

B) I recommend to use as less dependencies as possible. In this case, Google Enums is not necessarily needed. You could rewrite the code in the following way, which is also more compact:

```
private String getLabel(IWell well) {

	LabelSetting labelSetting = getLabelSetting();
	switch(labelSetting) {
		case SAMPLENAME:
			return well.getSampleId();
		case COORDINATE:
			return well.getPosition().toString();
		case COORDINATE_SAMPLENAME:
		default:
			return well.getPosition().toString() + ": " + well.getSampleId();
	}
}

private LabelSetting getLabelSetting() {

	try {
		return LabelSetting.valueOf(preferenceStore.getString(PreferenceConstants.P_PCR_REFERENCE_LABEL));
	} catch(Exception e) {
		return LabelSetting.COORDINATE_SAMPLENAME;
	}
}

```

Mailaender committed:
Add a configuration option on how to label PCR data.

Mailaender committed:
Put the settings button on the end for consistency.

Mailaender commented:
Rebased, removed Google Guave, initialized the setting and put the settings button on the right.
